dma: hda: dwdma: adsp: ace: Extend DMA api functionality

softwarecki opened the pull request:
This PR adds:
- new fields containing linear link position to the dma_status structure. 
- reading of the linear link position in the status function in adsp gpdma driver. 
- new get_attribute function to the dma api.
- get attribute function in dwdma and alh drivers.

PR at SOF side: https://github.com/thesofproject/sof/pull/6538

softwarecki committed:
dma: hda: adsp: Add support for get_attribute

Added the new intel_adsp_hda_dma_get_attribute function that
returns dma controller attributes.

Signed-off-by: Adrian Warecki <adrian.warecki@intel.com>

softwarecki committed:
dma: gpdma: Read linear link position in status function

Added reading of the linear link position in the status function.

Signed-off-by: Adrian Warecki <adrian.warecki@intel.com>

softwarecki committed:
dma: Extend dma status api to get total copied data

Added a new field containing total copied data to the
dma_status structure.

Signed-off-by: Adrian Warecki <adrian.warecki@intel.com>

softwarecki committed:
dma: Extend api to allow get controller attributes

Added a new get_attribute function to the dma driver api.

Signed-off-by: Adrian Warecki <adrian.warecki@intel.com>

abonislawski commented on drivers/dma/dma_intel_adsp_hda.c:
```diff
@@ -265,3 +265,33 @@ int intel_adsp_hda_dma_init(const struct device *dev)
+/* DGMBS align value */
+#define HDA_DMA_BUFFER_ALIGNMENT		0x20
+#define HDA_DMA_COPY_ALIGNMENT			0x20
+#define HDA_DMA_BUFFER_ADDRESS_ALIGNMENT	0x80
```
This is already in dtsi: dma-buf-alignment

abonislawski commented on drivers/dma/dma_intel_adsp_hda.c:
```diff
@@ -265,3 +265,33 @@ int intel_adsp_hda_dma_init(const struct device *dev)
 }
+
+/* DGMBS align value */
+#define HDA_DMA_BUFFER_ALIGNMENT		0x20
```
@teburd we should move all of this to dtsi or maybe remove every alignment from dtsi and stick with c defines?
The current mix is probably not looking good

teburd replied:
The nice part of having these things in device tree, if you use static DEVICE_DT_GET() style macros to get the device, you can also statically get the attributes. And statically setup buffers using those attributes. That's really nice from my perspective.

Those properties aren't available dynamically though. It makes sense, for sof especially, to have the get_attribute() matching up with the existing DMA API there. This could easily return the attributes assigned from DT instead of C defines and you maybe get the best of both.

teburd replied:
I'm not sure I agree that you would need to store them in a structure, you can get the DT properties just like you have here in the function body without storing them somewhere.

The bigger issue seems to be needing to replicate the properties for each node which I agree is a chore. I'm discussing with @galak to sort out how to avoid that particular issue and how best to do it.

We need these properties of the hardware to be accessible at compile time. While SoF uses dynamic memory allocation in its dai-zephyr.c code, Zephyr tests do not. Generally speaking, Zephyr code and the code that builds upon it wants to know things at compile time which makes sense. The information is there in device tree to do so.

I agree that dma_get_attribute is useful and should be added. What I don't like is that these hardware attributes are now tucked away in a .c file unusable by test code that exists today in Zephyr.

teburd commented on include/zephyr/drivers/dma.h:
```diff
@@ -226,6 +234,8 @@ struct dma_status {
 	uint32_t free;
 	uint32_t write_position;
 	uint32_t read_position;
+	uint32_t llp_l;
```
The llp_l/llp_u status members are pretty specific to SOF, how it uses DMA, and Intel hardware. These should be ifdef'ed out with something like DMA_STATUS_LLP with a description of when its used.

kv2019i replied:
I have the same feeling that this is too audio specific. Can we use some more generic terms? And please define (in the commit and struct documentation) how this is different from "read_position" and "write_position".

teburd replied:
Something like

uint64_t copied;

perhaps? Implying the total number of something (audio samples/words/bytes, whatever is configured...) copied is probably useful, and would not require an ifdef.

kv2019i commented on include/zephyr/drivers/dma.h:
```diff
@@ -66,6 +66,14 @@ enum dma_channel_filter {
+	DMA_ATTR_BUFFER_ALIGNMENT,
+	DMA_ATTR_COPY_ALIGNMENT,
+	DMA_ATTR_BUFFER_ADDRESS_ALIGNMENT,
+	DMA_ATTR_BUFFER_PERIOD_COUNT,
```
Is "period" a generic DMA term? This has established meaning in audio domain, but not sure if it's correct to use in generic DMA driver interface.

abonislawski replied:
Thats right, period should be app (sof) specific. I can see we already have DW_DMA_BUFFER_PERIOD_COUNT defined in zephyr dma_dw_common.h without usage, probably should be removed

kv2019i reviewed (changes requested):
Late to the review, but I'll echo the definitions in generic dma.h need to be generic and not specific to audio (and use HDA terms like "link" unless clearly explained).

marcinszkudlinski commented on include/zephyr/drivers/dma.h:
```diff
@@ -582,6 +595,30 @@ static inline int dma_get_status(const struct device *dev, uint32_t channel,
+/**
+ * @brief get attribute of a dma controller
+ *
+ * Implementations must check the validity of the type passed in and
```
Be more verbose pls, the comment gives no clue what the procedure really does - what "an attribute", etc.???

"a device specific static or runtime attribute like required alignment size" or similar

softwarecki committed:
dma: dts: Rename of the dma_buf_alignment to dma-buf-addr-alignment

Renamed the dma-buf-alignment field to a more explicit
and descriptive name dma-buf-addr-alignment.

Signed-off-by: Adrian Warecki <adrian.warecki@intel.com>

softwarecki committed:
dma: dts: Add support for dma-copy-alignment and dma-buf-size-alignment

dma-buf-size-alignment: Buffer size alignment required by the DMA
controller.

dma-copy-alignment: Minimal chunk of data possible to be copied
by the controller.

Signed-off-by: Adrian Warecki <adrian.warecki@intel.com>

softwarecki committed:
dma: dts: hda: Add controller attributes to DT

Added to the device tree values of the dma-copy-alignment
and dma-buf-size-alignment attributes.

Signed-off-by: Adrian Warecki <adrian.warecki@intel.com>

softwarecki committed:
dma: dwdma: Remove unused defines

There was no reference to the DW_DMA_BUFFER_PERIOD_COUNT,
so it was removed.

Signed-off-by: Adrian Warecki <adrian.warecki@intel.com>

softwarecki committed:
dma: dts: gpdma: Add controller attributes to DT

Added to the device tree values of the dma-copy-alignment
and dma-buf-size-alignment attributes.

Signed-off-by: Adrian Warecki <adrian.warecki@intel.com>

softwarecki committed:
dma: adsp: gpdma: Add support for get_attribute

Added the new intel_adsp_gpdma_get_attribute function that returns
dma controller attributes.

Signed-off-by: Adrian Warecki <adrian.warecki@intel.com>

teburd reviewed (commented):
This looks much better, passing CI is a +1 from me.

kv2019i reviewed (dismissed):
Thanks, looks better now!

nashif committed:
manifest: update sof module

Update to a recent commit using new DMA APIs implemented in Zephyr.

Signed-off-by: Anas Nashif <anas.nashif@intel.com>
